js/eventloop.go

```diff
@@ -0,0 +1,165 @@
+// are supposed to be independent and any work started in them needs to finish,
+// but also they need to end when all the instructions are done.
+// Additionally because of this on any error while the event loop will exit it's
+// required to wait on the event loop to be empty before the execution can  continue.
+type eventLoop struct {
+	lock          sync.Mutex
+	queue         []func() error
+	wakeupCh      chan struct{} // maybe use sync.Cond ?
+	reservedCount int
+
+	// pendingPromiseRejections are rejected promises with no handler,
+	// if there is something in this map at an end of an event loop then it will exit with an error.
+	// It's similar to what Deno and Node do.
+	pendingPromiseRejections map[*goja.Promise]struct{}
+}
+
+// newEventLoop returns a new event loop with a few helpers attach on it:
+// - adding setTimeout javascript implementation
+// - reporting (and aborting on) unhandled promise rejections
+func newEventLoop(ctx context.Context, rt *goja.Runtime) *eventLoop {
+	e := &eventLoop{
+		wakeupCh:                 make(chan struct{}, 1),
+		pendingPromiseRejections: make(map[*goja.Promise]struct{}),
+	}
+	rt.SetPromiseRejectionTracker(e.promiseRejectionTracker)
+	e.addSetTimeout(ctx, rt)
+
+	return e
+}
+
+func (e *eventLoop) wakeup() {
+	select {
+	case e.wakeupCh <- struct{}{}:
+	default:
+	}
+}
+
+// reserve "reserves" a spot on the loop, preventing it from returning/finishing. The returning function will queue it's
+// argument and wakeup the loop if needed.
+// If the eventLoop has since stopped it will not be execeted
+// This function *must* be called from within running on the event loop, but it's result can be called from anywhere
+// TODO better name
+func (e *eventLoop) reserve() func(func() error) {
+	e.lock.Lock()
+	e.reservedCount++
+	e.lock.Unlock()
+
+	return func(f func() error) {
+		e.lock.Lock()
+		e.queue = append(e.queue, f)
+		e.reservedCount--
+		e.lock.Unlock()
+		e.wakeup()
+	}
+}
+
+func (e *eventLoop) promiseRejectionTracker(p *goja.Promise, op goja.PromiseRejectionOperation) {
+	// No locking here as the goja runtime will call this synchrnously
+	// Read Notes on https://tc39.es/ecma262/#sec-host-promise-rejection-tracker
+	if op == goja.PromiseRejectionReject {
+		e.pendingPromiseRejections[p] = struct{}{}
+	} else { // goja.PromiseRejectionHandle so a promise that was previously rejected without handler now got one
+		delete(e.pendingPromiseRejections, p)
+	}
+}
+
+func (e *eventLoop) popAll() (queue []func() error, reserved bool) {
+	e.lock.Lock()
+	queue = e.queue
+	e.queue = make([]func() error, 0, len(queue))
+	reserved = e.reservedCount != 0
+	e.lock.Unlock()
+	return
+}
+
+// start will run the event loop until it's empty and there are no reserved spots
+// or a queued function returns an error. The provided function will be the first thing executed.
+// After start returns the event loop should not be reused if an error was returend
+func (e *eventLoop) start(f func() error) error {
```

Comment:
I would like for the parameter `f` to have a more explicit name if possible. It was not immediately clear after reading the code, nor the method's document, what it was actually used for, nor how or why it should be used 🙇🏻

Reply:
better in [5c5878a](https://github.com/grafana/k6/pull/2228/commits/5c5878acae7759bb99bd10f337d46cc5b16d1e81)